compiler/rustc_abi/src/layout.rs

```diff
@@ -8,17 +8,18 @@ use rand_xoshiro::Xoshiro128StarStar;
 
 use tracing::debug;
 
-// Invert a bijective mapping, i.e. `invert(map)[y] = x` if `map[x] = y`.
-// This is used to go between `memory_index` (source field order to memory order)
-// and `inverse_memory_index` (memory order to source field order).
-// See also `FieldsShape::Arbitrary::memory_index` for more details.
-// FIXME(eddyb) build a better abstraction for permutations, if possible.
-fn invert_mapping(map: &[u32]) -> Vec<u32> {
-    let mut inverse = vec![0; map.len()];
-    for i in 0..map.len() {
-        inverse[map[i] as usize] = i as u32;
-    }
-    inverse
+fn transmute_index_vec(inv: Vec<FieldIdx>) -> Vec<u32> {
+    assert_eq!(std::alloc::Layout::new::<u32>(), std::alloc::Layout::new::<FieldIdx>());
+
+    let mut inv = std::mem::ManuallyDrop::new(inv);
+    let ptr = inv.as_mut_ptr().cast::<u32>();
+    let len = inv.len();
+    let cap = inv.capacity();
+
+    // SAFETY: `u32` allows more bit patterns than `FieldIdx`,
+    // and they have exactly the same size and alignment.
+    // `inv` isn't going to be dropped, so no double-free.
+    unsafe { Vec::from_raw_parts(ptr, len, cap) }
 }
```

Comment:
This file is already quite complex, so I don't think it's worth going to such effort to transmute this vec, vs. a naive `inv.into_iter().map(|i| i.as_u32()).collect()`, unless there is a measurable perf improvement. (Layout code tends not to be perf sensitive IME--only its behavior has perf implications.)

It's likely that the collect-in-place specialization will make the naive implementation equivalent anyways: https://godbolt.org/z/xdKE6zzar

Reply:
Ah, I assumed it was important since it wasn't just always inverting the mapping.

The magic in-place specialization is a good point; will switch.